File: pyguymer3/geo/_buffer_points_crudely.py
#!/usr/bin/env python3

# Define function ...
def _buffer_points_crudely(points1, dist, nang, /, *, eps = 1.0e-12, nmax = 100, ramLimit = 1073741824):
    """Buffer some points

    This function reads in an array of coordinates (in degrees) that exist on
    the surface of the Earth and returns an array of coordinates that are the
    rings around the input coordinates buffered by a constant distance (in
    metres).

    Parameters
    ----------
    points1 : numpy.ndarray
        the (npoint, 2) array of (lon,lat) coordinates (in degrees)
    dist : float
        the distance to buffer the (lon,lat) coordinates by (in metres)
    nang : int
        the number of angles around the (lon,lat) coordinates that are
        calculated when buffering (must be odd; must be ≥ 9)
    ramLimit : int, optional
        the maximum RAM usage of each "large" array, in bytes

    Returns
    -------
    points2 : numpy.ndarray
        the (npoint, nang, 2) array of (lon,lat) coordinates around the
        (lon,lat) coordinates (in degrees)

    Notes
    -----
    Copyright 2017 Thomas Guymer [1]_

    References
    ----------
    .. [1] PyGuymer3, https://github.com/Guymer/PyGuymer3
    """

    # Import special modules ...
    try:
        import numpy
    except:
        raise Exception("\"numpy\" is not installed; run \"pip install --user numpy\"") from None

    # Import sub-functions ...
    from .calc_loc_from_loc_and_bearing_and_dist import calc_loc_from_loc_and_bearing_and_dist

    # Create short-hand ...
    npoint = points1.shape[0]

    # Check array size ...
    if npoint * nang * 2 * 8 > ramLimit:
        raise Exception(f"\"points2\" is going to be {npoint * nang * 2 * 8:,d} bytes, which is larger than {ramLimit:,d} bytes") from None

    # Initialize array ...
    points2 = numpy.zeros((npoint, nang, 2), dtype = numpy.float64)             # [°]

    # Loop over angles ...
    # NOTE: The first and last angles will *always* be exactly North (therefore,
    #       use that as a check later on).
    # NOTE: The middle angle will *always* be exactly South (therefore, use that
    #       as a check later on).
    # NOTE: The most two subsequent points can be apart is ~45° (with nang ≥ 9).
    for iang in range(nang - 1):
        # Calculate initial angle ...
        ang1 = 360.0 * float(nang - 1 - iang) / float(nang - 1)                 # [°]

        # Loop over points ...
        for ipoint in range(npoint):
            # Calculate the ring coordinates and add them to the array ...
            points2[ipoint, iang, 0], points2[ipoint, iang, 1], _ = calc_loc_from_loc_and_bearing_and_dist(
                points1[ipoint, 0],
                points1[ipoint, 1],
                ang1 % 360.0,
                dist,
                 eps = eps,
                nmax = nmax,
            )                                                                   # [°], [°]

    # Loop over points ...
    for ipoint in range(npoint):
        # Force the last point to be the same as the first point ...
        points2[ipoint, -1, :] = points2[ipoint, 0, :]                          # [°]

    # Return answer ...
    return points2

File: pyguymer3/f90/src/buffer_points_crudely.f90
SUBROUTINE buffer_points_crudely(points1, dist, nang, npoint, points2)
    ! Import standard modules ...
    USE ISO_C_BINDING

    ! Import my modules ...
    USE mod_safe, ONLY: sub_calc_loc_from_loc_and_bearing_and_dist

    IMPLICIT NONE

    ! Declare inputs/outputs ...
    INTEGER(kind = C_LONG_LONG), INTENT(in)                                     :: nang
    INTEGER(kind = C_LONG_LONG), INTENT(in)                                     :: npoint
    REAL(kind = C_DOUBLE), INTENT(in)                                           :: dist
    REAL(kind = C_DOUBLE), DIMENSION(npoint, 2), INTENT(in)                     :: points1
    REAL(kind = C_DOUBLE), DIMENSION(npoint, nang, 2), INTENT(out)              :: points2

    ! Declare internal variables ...
    INTEGER(kind = C_LONG_LONG)                                                 :: iang
    INTEGER(kind = C_LONG_LONG)                                                 :: ipoint
    REAL(kind = C_DOUBLE)                                                       :: ang1
    REAL(kind = C_DOUBLE)                                                       :: ang2

    !$omp parallel                                                              &
    !$omp default(none)                                                         &
    !$omp private(ang1)                                                         &
    !$omp private(ang2)                                                         &
    !$omp private(iang)                                                         &
    !$omp private(ipoint)                                                       &
    !$omp shared(dist)                                                          &
    !$omp shared(nang)                                                          &
    !$omp shared(npoint)                                                        &
    !$omp shared(points1)                                                       &
    !$omp shared(points2)
        !$omp do                                                                &
        !$omp schedule(dynamic)
            ! Loop over angles ...
            ! NOTE: The first and last angles will *always* be exactly North
            !       (therefore, use that as a check later on).
            ! NOTE: The middle angle will *always* be exactly South (therefore,
            !       use that as a check later on).
            ! NOTE: The most two subsequent points can be apart is ~45° (with
            !       nang ≥ 9).
            DO iang = 1_C_LONG_LONG, nang - 1_C_LONG_LONG
                ! Calculate initial angle ...
                ang1 = 360.0e0_C_DOUBLE * REAL(iang - 1_C_LONG_LONG, kind = C_DOUBLE) / REAL(nang - 1_C_LONG_LONG, kind = C_DOUBLE)

                ! Loop over points ...
                DO ipoint = 1_C_LONG_LONG, npoint
                    ! Calculate the ring coordinates and add them to the array ...
                    CALL sub_calc_loc_from_loc_and_bearing_and_dist(points1(ipoint, 1), points1(ipoint, 2), ang1, dist, points2(ipoint, iang, 1), points2(ipoint, iang, 2), ang2)
                END DO
            END DO
        !$omp end do
    !$omp end parallel

    !$omp parallel                                                              &
    !$omp default(none)                                                         &
    !$omp private(ipoint)                                                       &
    !$omp shared(nang)                                                          &
    !$omp shared(npoint)                                                        &
    !$omp shared(points2)
        !$omp do                                                                &
        !$omp schedule(dynamic)
            ! Loop over points ...
            DO ipoint = 1_C_LONG_LONG, npoint
                ! Force the last point to be the same as the first point ...
                points2(ipoint, nang, 1) = points2(ipoint, 1, 1)
                points2(ipoint, nang, 2) = points2(ipoint, 1, 2)
            END DO
        !$omp end do
    !$omp end parallel
END SUBROUTINE buffer_points_crudely
